iroha_p2p/src/peer.rs

```diff
@@ -29,422 +27,701 @@ use ursa::{
+            w,
+            self.cipher.unwrap(),
+            self.network,
+            false,
+            sender,
+        );
+        let stream = Peer::<T, K, E>::stream(r, receiver);
+
+        let peer = peer.preinit();
+        let addr = peer.address.clone();
+
+        // spawning task
+        tokio::spawn(async move {
+            stream.for_each(|it| async { addr.do_send(it).await }).await;
+        });
+
+        peer.start().await
     }
 }
 
-impl<T, K, E> Debug for Peer<T, K, E>
+#[async_trait::async_trait]
+impl<T, K, E> Actor for FromPeer<T, K, E>
 where
-    T: Encode + Decode + Send + Clone + 'static,
+    T: Debug + Encode + Decode + Send + Sync + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("Peer")
-            .field("id", &self.id)
-            .field("state", &self.state)
-            .finish()
+    async fn on_start(&mut self, ctx: &mut Context<Self>) {
+        info!(addr = ?self.id.address, "Connecting and doing handshake");
+        ctx.notify(ContinueHandshake);
+    }
+
+    async fn on_stop(self, _ctx: Context<Self>) {
+        self.into_peer().await;
     }
 }
 
 #[async_trait::async_trait]
-impl<T, K, E> Actor for Peer<T, K, E>
+impl<T, K, E> ContextHandler<ContinueHandshake> for FromPeer<T, K, E>
 where
-    T: Encode + Decode + Send + Sync + Clone + 'static,
+    T: Debug + Encode + Decode + Send + Sync + Clone + 'static,
     K: KeyExchangeScheme + Send + 'static,
     E: Encryptor + Send + 'static,
 {
-    async fn on_start(&mut self, ctx: &mut Context<Self>) {
-        debug!(
-            "Starting actor for connection with peer {}",
-            &self.id.address
-        );
-        let mut count = 0;
-        while self.state != State::Ready {
-            if let Err(e) = self.handshake().await {
-                warn!(
-                    "Error connecting to peer {} in state {}. {}",
-                    &self.id.address, &self.state, e
-                );
-                if count < MAX_CONNECTION_RETRY_COUNT {
-                    count += 1;
-                    let delay = CONNECTION_RETRY_PERIOD * count;
-                    debug!(%e, "Error connecting to peer {}, waiting {}ms before retry.", &self.id.address, delay);
-                    time::sleep(Duration::from_millis(delay)).await;
-                    continue;
+    type Result = ();
+    async fn handle(
+        &mut self,
+        ctx: &mut Context<Self>,
+        ContinueHandshake: ContinueHandshake,
+    ) -> Self::Result {
+        if self.state == FromState::Ready {
+            ctx.stop_now();
+            return;
+        }
+
+        if let Err(err) = self.handshake().await {
+            info!(addr=?self.id.address, state=?self.state, %err, "Error connecting to peer");
+            ctx.stop_now();
+            return;
+        }
+        ctx.notify(ContinueHandshake);
```

Comment:
Doing simple loop through actors/channels? Why?

Reply:
In order to be able to stop it, during handshake

Reply:
Simple `break` will not suffice?

Reply:
I mean to stop by network